### PR TITLE
983: Adding timeout in waitRules

### LINKED ISSUE #983: Enhance custom waiting behavior.
**Describe the problem/challenge you have**
There are certain scenarios where a custom resource moves to the `FAILED` state before coming to the `SUCCESSFUL` state. e.g.  if a resource has external dependency as part of its readiness criteria, and this external dependency itself takes time to come, then resource moves to FAILED state and `kapp deploy` exits with error. But once the external service comes up, custom resource readiness criteria is fulfilled and it moves to the SUCCESSFUL state. 

**Describe the solution you'd like**
I would like `kapp deploy` to not fail aggressively and rather it should fail only if the app has been in `FAILED` state for lets say 5 min. 

**Anything else you would like to add:**
[Additional information that will assist in solving the issue.]

---
Vote on this request

This is an invitation to the community to vote on issues, to help us prioritize our backlog. Use the "smiley face" up to the right of this comment to vote.

👍 "I would like to see this addressed as soon as possible"
👎 "There are other more important things to focus on right now"

We are also happy to receive and review Pull Requests if you want to help working on this issue.

### PR DESCRIPTION
**Pull request by rohitagg2020**
Sample Output

```
$ kapp deploy -a simpleapp -f ~/config/pod-config.yml
Target cluster 'https://127.0.0.1:58277' (nodes: kind-control-plane)

Changes

Namespace  Name   Kind  Age  Op      Op st.  Wait to    Rs  Ri
default    nginx  Pod   -    create  -       reconcile  -   -

Op:      1 create, 0 delete, 0 update, 0 noop, 0 exists
Wait to: 1 reconcile, 0 delete, 0 noop

Continue? [yN]: y

11:08:45AM: ---- applying 1 changes [0/1 done] ----
11:08:46AM: create pod/nginx (v1) namespace: default
11:08:46AM: ---- waiting on 1 changes [0/1 done] ----
11:08:46AM: ongoing: reconcile pod/nginx (v1) namespace: default
11:08:46AM:  ^ ContainersReady: ContainersNotReady (message: containers with unready status: [nginx])
11:09:16AM: fail: reconcile pod/nginx (v1) namespace: default
11:09:16AM:  ^ Encountered failure condition ContainersReady == False: ContainersNotReady (message: containers with unready status: [nginx]) continuously for 30s duration

kapp: Error: waiting on reconcile pod/nginx (v1) namespace: default:
  Finished unsuccessfully (Encountered failure condition ContainersReady == False: ContainersNotReady (message: containers with unready status: [nginx]) continuously for 30s duration)
```

#### Which issue(s) this PR fixes:

Fixes #

#### Does this PR introduce a user-facing change?

```release-note

```

Fixes #983 
##### Review Checklist:

- [ ] Follows the [developer guidelines](https://carvel.dev/shared/docs/latest/development_guidelines/)
- [ ] Relevant tests are added or updated
- [ ] Relevant docs in this repo added or updated
- [ ] Relevant carvel.dev docs added or updated in a separate PR and there's
  a link to that PR
- [ ] Code is at least as readable and maintainable as it was before this
  change

#### Additional documentation e.g., Proposal, usage docs, etc.:

```docs

```

### CODE REVIEW

File: pkg/kapp/resourcesmisc/custom_waiting_resource.go [hunk 1]
```diff
@@ -132,3 +161,20 @@ func (s CustomWaitingResource) IsDoneApplying() DoneApplyState {
...
+func (s CustomWaitingResource) hasTimeoutOccurred(timeout string, key string) bool {
+	expiryTime, found := timeoutMap.Load(key)
+	if found {
+		return time.Now().Sub(expiryTime.(time.Time)) > 0
+	}
+	timeoutMap.Store(key, time.Now().Add(parseDuration(timeout)))
+	return false
+}
+
+func parseDuration(str string) time.Duration {
```
Comment (100mik) [inline pkg/kapp/resourcesmisc/custom_waiting_resource.go]: Let's do this inline?

File: pkg/kapp/resourcesmisc/custom_waiting_resource.go [hunk 2]
```diff
@@ -81,16 +85,41 @@ func (s CustomWaitingResource) IsDoneApplying() DoneApplyState {
...
+					}
+					return DoneApplyState{Done: false, Message: fmt.Sprintf(
+						"%s: %s (message: %s)",
+						cond.Type, cond.Reason, cond.Message)}
+				}
+			}
+			if !isTimeOutConditionPresent {
+				timeoutMap.Delete(fmt.Sprintf("%s.%s", s.resource.Namespace(), s.resource.Name()))
+			}
+		} else {
```
Comment (100mik) [inline pkg/kapp/resourcesmisc/custom_waiting_resource.go]: nit: Let's return early and drop the else block?

File: pkg/kapp/resourcesmisc/custom_waiting_resource.go (hunk 1 above)
Comment (rohitagg2020) [reply pkg/kapp/resourcesmisc/custom_waiting_resource.go]: Adopted

File: pkg/kapp/resourcesmisc/custom_waiting_resource.go (hunk 2 above)
Comment (rohitagg2020) [reply pkg/kapp/resourcesmisc/custom_waiting_resource.go]: Adopted

**Review by praveenrewar — APPROVED** [review]
LGTM!

**Comment by 100mik** [comment]
We do need to document this behaviour

**Pull request merged**